consoles/video_stream.pm

```diff
@@ -0,0 +1,350 @@
+# Copyright 2021 Marek Marczykowski-Górecki
+# Copyright 2021 SUSE LLC
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+package consoles::video_stream;
+
+use Mojo::Base -strict, -signatures;
+
+use base 'consoles::network_console';
+
+use List::Util 'max';
+use Time::HiRes qw(usleep);
+
+use Try::Tiny;
+use testapi qw(get_var DEFAULT_MAX_INTERVAL);
+use bmwqemu;
+
+# speed limit: 30 keys per second
+use constant STREAM_TYPING_LIMIT_DEFAULT => 30;
+
+use constant DV_TIMINGS_CHECK_INTERVAL => 3;
+
+use constant STALL_THRESHOLD => 4;
+
+my $CHARMAP = {
+    "\t" => 'tab',
+    "\n" => 'ret',
+    "\b" => 'backspace',
+    "\e" => 'esc',
+    " " => 'space',
+};
+
+sub screen ($self, @) {
+    return $self;
+}
+
+sub disable_video ($self) {
+    my $ret = 0;
+    if ($self->{ffmpeg}) {
+        kill(TERM => $self->{ffmpegpid});
+        close($self->{ffmpeg});
+        $self->{ffmpeg} = undef;
+        $ret = waitpid($self->{ffmpegpid}, 0);
+        $self->{ffmpegpid} = undef;
+    }
+    return $ret;
+}
+
+sub disable ($self, @) {
+    my $ret = $self->disable_video;
+    if ($self->{input_pipe}) {
+        close($self->{input_pipe});
+        waitpid($self->{inputpid}, 0);
+    }
+    return $ret;
+}
+
+sub _v4l2_ctl ($device, $cmd) {
+    my @cmd = ("v4l2-ctl", "--device", $device, "--concise");
+    push(@cmd, split(/ /, $cmd));
+    my $pipe;
+    my $pid = open($pipe, '-|', @cmd) or return undef;
+    $pipe->read(my $str, 50);
+    my $ret = waitpid($pid, 0);
+    if ($ret > 0 && $? == 0) {
+        # remove header and whitespaces
+        $str =~ s/DV timings://;
+        $str =~ s/^\s+|\s+$//g;
+        return $str;
+    }
+    return undef;
+}
+
+sub connect_remote ($self, $args) {
+    $self->{_last_update_received} = 0;
+
+    if ($args->{url} =~ m/^\/dev\/video/) {
+        if ($args->{edid}) {
+            my $ret = _v4l2_ctl($args->{url}, "--set-edid $args->{edid}");
+            die "Failed to set EDID" unless defined $ret;
+        }
+
+        my $timings = _v4l2_ctl($args->{url}, '--get-dv-timings');
+        if ($timings) {
+            if ($timings ne "0x0pnan") {
+                $self->{dv_timings} = $timings;
+            } else {
+                $self->{dv_timings} = '';
+            }
+            $self->{dv_timings_supported} = 1;
+            $self->{dv_timings_last_check} = time;
+            bmwqemu::diag "Current DV timings: $timings";
+        } else {
+            $self->{dv_timings_supported} = 0;
+            bmwqemu::diag "DV timings not supported";
+        }
+    } else {
+        # applies to v4l only
+        $self->{dv_timings_supported} = 0;
+    }
+
+    bmwqemu::diag "Starting to receive video stream at $args->{url}";
+    $self->connect_remote_video($args->{url});
+
+    $self->connect_remote_input($args->{input_cmd}) if $args->{input_cmd};
+}
+
+sub _get_ffmpeg_cmd ($self, $url) {
+    my @cmd = ('/usr/bin/ffmpeg', '-loglevel', 'fatal', '-i', $url);
```

Comment:
Is hard-coding the absolute path required here? It would be nice to make this configurable (via `PATH` or a specific env variable).

Reply:
Apparently using just `ffmpeg` simply works (via `PATH`), so I changed this.